evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -1,34 +1,145 @@
+        hash32,
+        address,
+        hash32,
+        hash32,
+        rlp.sedes.big_endian_int,
+        rlp.sedes.binary,
+    ])
+    header_values = rlp.decode(header_bytes, sedes=sedes)
+    yield 'header', header_values
+    yield 'is_new_head', is_new_head
+    yield 'score', score
+
+
 class VMC(Contract):
 
-    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
+    logger = logging.getLogger("evm.chain.sharding.VMC")
+
+    # For handling logs filtering
+    # Event:
+    #   CollationAdded(indexed uint256 shard, bytes collationHeader, bool isNewHead, uint256 score)
+    COLLATION_ADDED_TOPIC = event_signature_to_log_topic(
+        "CollationAdded(int128,bytes4096,bool,int128)"
+    )
 
-    def __init__(self, *args, default_privkey, **kwargs):
+    shards = set()
+    # shard_id -> list
+    # older <---------------> newer
+    new_collation_added_logs = {}
+    # shard_id -> list
+    # newer <---------------> older
+    unchecked_collation_added_logs = {}
+    # shard_id -> score
+    current_checking_score = {}
+
+    def __init__(self, *args, log_handler, default_privkey, **kwargs):
+        self.log_handler = log_handler
         self.default_privkey = default_privkey
         self.default_sender_address = default_privkey.public_key.to_canonical_address()
         self.config = get_sharding_config()
         super().__init__(*args, **kwargs)
 
+    def init_shard_variables(self, shard_id):
+        self.shards.add(shard_id)
+        self.unchecked_collation_added_logs[shard_id] = []
+        self.new_collation_added_logs[shard_id] = []
+        self.current_checking_score[shard_id] = None
+
+    def ensure_shard_variables_initialied(self, shard_id):
+        if shard_id not in self.shards:
+            self.init_shard_variables(shard_id)
+
+    @to_tuple
+    def _get_new_logs(self, shard_id):
+        self.ensure_shard_variables_initialied(shard_id)
+        shard_id_topic_hex = encode_hex(shard_id.to_bytes(32, byteorder='big'))
+        new_logs = self.log_handler.get_new_logs(
+            address=self.address,
+            topics=[
+                encode_hex(self.COLLATION_ADDED_TOPIC),
+                shard_id_topic_hex,
+            ],
+        )
+        for log in new_logs:
+            yield parse_collation_added_data(log['data'])
+
+    def get_next_log(self, shard_id):
+        new_logs = self._get_new_logs(shard_id)
+        self.new_collation_added_logs[shard_id] += new_logs
+        if len(self.new_collation_added_logs[shard_id]) == 0:
+            raise NextLogUnavailable("No more next logs")
+        return self.new_collation_added_logs[shard_id].pop()
+
+    def fetch_candidate_head(self, shard_id):
+        # Try to return a log that has the score that we are checking for,
+        # checking in order of oldest to most recent.
+        self.ensure_shard_variables_initialied(shard_id)
+
+        for i in reversed(range(len(self.unchecked_collation_added_logs[shard_id]))):
```

Comment:
I think this is more readable as:

```python
# tuple(..) call is to force the iterator to fully evaluate so that the later `.pop(..)` doesn't cause problems.
unchecked_logs = tuple(reversed(enumerate(self.unchecked_collation_added_logs[shard_id])))
current_score = self.current_checking_score[shard_id]

for idx, log_entry in unchecked_logs:
    if log_entry['score'] == current_score:
        return self.unchecked_collation_added_logs[shard_id].pop(idx)
```

It's really just moving things around, but it removes one index based lookup, as well as storing the current score in a local variable to make the inner loop more legible and easier to follow.

Reply:
Thank you for the suggestion! It is way more readable. Will change it.